samples/swift/uidemo/ChatViewController.swift

```diff
@@ -0,0 +1,251 @@
+// to populate a collection view with chat messages. The relevant code
+// is in the call to `collectionViewDataSource.populateCellWithBlock`.
+//
+// All of the error handling in this controller is done with `fatalError`;
+// please don't copy paste it into your production code.
+class ChatViewController: UIViewController, UICollectionViewDelegateFlowLayout {
+  
+  private static let reuseIdentifier = "ChatCollectionViewCell"
+  
+  @IBOutlet private var collectionView: UICollectionView!
+  @IBOutlet private var textView: UITextView! {
+    didSet {
+      textView.layer.borderColor = UIColor.grayColor().colorWithAlphaComponent(0.5).CGColor
+      textView.layer.borderWidth = 1
+      textView.layer.cornerRadius = 8
+      textView.layer.masksToBounds = true
+    }
+  }
+  @IBOutlet private var sendButton: UIButton!
+  
+  // Used to shift view contents up when the keyboard appears.
+  @IBOutlet weak var bottomConstraint: NSLayoutConstraint!
+  
+  private let auth = FIRAuth.auth()
+  private let chatReference = FIRDatabase.database().reference().child("chats")
+  
+  private var collectionViewDataSource: FirebaseCollectionViewDataSource! = nil
+  
+  private var user: FIRUser?
+  private var query: FIRDatabaseQuery?
+  
+  static func fromStoryboard(storyboard: UIStoryboard = AppDelegate.mainStoryboard) -> ChatViewController {
+    return storyboard.instantiateViewControllerWithIdentifier("ChatViewController") as! ChatViewController
+  }
+  
+  override func viewDidLoad() {
+    super.viewDidLoad()
+    
+    self.collectionView.backgroundColor = UIColor.whiteColor()
+    self.collectionView.delegate = self
+    let layout = self.collectionView.collectionViewLayout as! UICollectionViewFlowLayout
+    layout.minimumInteritemSpacing = CGFloat.max
+    layout.minimumLineSpacing = 4
+    
+    self.auth?.addAuthStateDidChangeListener { (auth, user) in
+      self.user = user
+    }
+    
+    self.sendButton.addTarget(self, action: #selector(didTapSend), forControlEvents: .TouchUpInside)
+  }
+  
+  override func viewDidAppear(animated: Bool) {
+    super.viewDidAppear(animated)
+    
+    self.auth?.signInAnonymouslyWithCompletion { (user, error) in
+      if let error = error {
+        fatalError("Sign in failed: \(error.localizedDescription)")
+      }
+      
+      self.query = self.chatReference.queryLimitedToLast(50)
+      
+      self.collectionViewDataSource = FirebaseCollectionViewDataSource(query: self.query!,
+        prototypeReuseIdentifier: ChatViewController.reuseIdentifier,
+        view: self.collectionView)
+      
+      // The initializer called above--though it takes a collection view--
+      // doesn't actually set the collection view's data source, so if
+      // we don't set it before trying to populate our view our app will crash.
+      self.collectionView.dataSource = self.collectionViewDataSource
+      
+      self.collectionViewDataSource.populateCellWithBlock { (anyCell, data) in
```

Comment:
The collection view inner block is also quite long and kind of distracts. Consider moving this setup to a separate function and sticking it lower in the controller (as its largely app specific)